text/primitive-projections.md

```diff
@@ -0,0 +1,70 @@
+  - Advantages: a simple model; no need to identify `Proj (f,x)` and `App (Const f, params, x)` in conversion
+  - Drawbacks: maybe a few incompatibilities due to the dropping of arguments?
+
+Related: [#14084](https://github.com/coq/coq/pull/14084) (`match goal` involving primitive projections).
+
+## Issue 2: the relative status of "folded" vs "unfolded" `Proj` node
+
+*Design choice*: The "folded" vs "unfolded" boolean is dropped. Unfolding a `Proj` node is a no-op. In scenario 2.1 of issue 1, unfolding a `Const denoting a projection reveals a `Proj` node. In scenario 2.2, it is a no-op.
+
+See associated [Zulip discussion](https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs.20.26.20plugin.20devs/topic/Primitive.20Projection.20mode).
+
+## Issue 3: the status of `match` and destructing `let` in "primitive" record types
+ 
+See e.g. [#13913](https://github.com/coq/coq/pull/13913).
+
+*Design choice 1*: keep `match` nodes as they are in primitive record types; they will behave correctly for reduction (that is like their expansion into projections); to deal with them in the conversion, it is enough to extend the conversion with eta-expansions of `match` into projections.
+
+## Issue 4: The status of the `t.(f args)` notation as an alternative notation for any kind of application.
+
+See e.g. [#11332](https://github.com/coq/coq/pull/11332).
+
+*Design choice 1*: Accept the notation `t.(f args)` as a shorthand for `f args t` even when `f` is not a projection even if this is lost at printing time.
+  - Advantages: nothing more to do
+  - Drawbacks: no printing
+
+*Design choice 2*: Give the ability to declare a table of "methods" supporting the `t.(f args)` notation. Only such registered method can use the notation. The notation is used for printing too. Internally, it is an `App` node.
+
+*Orthogonal design choice*: same as design choices 1 and 2 but with a different notation, e.g `t.{f args}`.
+
+## Issue 5: The representation of projections in "positive" types. Can they take benefit of a compact representation?
+
+*Design choice 1*: We don't make a difference between positive and negative types. Both supports a `Case` and a `Proj` node which behave anyway the same wrt reduction even when we think at them as syntactic sugar. We add eta-rule for `Case` in the conversion, which are the current rules for expansion into `Proj`. The only difference is that `Proj` is mentally thought as in normal form for a negative type and as a short-hand for a `match` for a positive type. As for `Case`, it can either be mentally thought as a short-hand for its let-proj-expansion for a negative type or explicitly expanded in `cbn` and `lazy`, while `Case` for a positive type is in normal form.
```

Comment:
Keep in mind that if the record has eta then 
~~~coq
Record R := r { A : Type; B : Type }.
Check fun x (f : match x with r a b => a -> b end) a b => f a b.
~~~
must somehow reduce the match on a variable. ie we cannot have Case be a normal form when it is on a type which has eta.

Conversely we can have projections with omitted parameters for types without eta, this is already the case for some coinductives. AFAIK it would work fine to allow both primitive projections and primitive match for non-eta inductives, they just wouldn't be convertible with each other.

Reply:
> ~~~coq
> Record R := r { A : Type; B : Type }.
> Check fun x (f : match x with r a b => a -> b end) a b => f a b.
> ~~~
> must somehow reduce the match on a variable.

Ah, interesting, this is a place where reduction rather than conversion is needed. So, something special to take care of `match` commutation would have to be done here.

> Conversely we can have projections with omitted parameters for types without eta, this is already the case for some coinductives. AFAIK it would work fine to allow both primitive projections and primitive match for non-eta inductives, they just wouldn't be convertible with each other.

Then, is the equational theory of record types without eta but with projections and match clear to you? Is it just about removing `t` ≣ `(fst t, snd t)` but keeping the commutation `match t as z return P with (x,y) => v end` ≣ `v[x:=fst t][y:=snd t]`?

Reply:
>Then, is the equational theory of record types without eta but with projections and match clear to you? Is it just about removing t ≣ (fst t, snd t) but keeping the commutation match t as z return P with (x,y) => v end ≣ v[x:=fst t][y:=snd t]?

No, this commutation is not well-typed without eta.

Reply:
Ah, right. Do we then need baclofen-style restrictions on the form of `P`?